### ISSUE TITLE
File upload must be in the root of the query

### ISSUE THREAD

**Issue by ngdang95** [apollographql/apollo-ios#961]
As the notes mentioned in the documentation site, [uploading files section](https://www.apollographql.com/docs/ios/mutations/#uploading-files), has one point:

> Due to some limitations around the spec, whatever the file is being added for should be at the root of your GraphQL query.
So if you have something like:    
`mutation AvatarUpload($userID: GraphQLID!, $file: Upload!) {
  id
}`
it will work.

But if we wrap it in a object like this:

`mutation AvatarUpload($avatarObject: AvatarObject!) {
  id
}`
Assump `AvatarObject` has **userID: GraphQLID** and **file: Upload**
It won't.

I also noticed that this one was resolved on the Android side.

So I'm really want to know why we couldn't do that on iOS? What are the limitations here?
Thank you 🤗

**Comment by designatednerd** [comment]
Honestly, it's because I haven't had time to get it to search recursively through a potentially infinite size of dictionary to replace a variable value with `null` in order to get the server to pick things up. 

[This is basically where things are happening](https://github.com/apollographql/apollo-ios/blob/master/Sources/Apollo/RequestCreator.swift#L108) - because of the way [the GraphQL multipart-form request spec](https://github.com/jaydenseric/graphql-multipart-request-spec) expects to find `null` in the variables array for variables that are represented by parts of the multipart-form data. Doing it at the root of the query makes this significantly simpler - in order to not do it at the root of the query you'd need to be able to search through the entire dictionary, then also make sure you're not replacing a variable with the same name but the exact variable you're trying to put in. This was basically a "we can get this working with a restriction" decision because getting it working perfectly was going to take a lot more time. 

I'd love to find a more flexible solution to this, but I've already got too many fish frying as is. Would love to see a PR with tests that makes this easier to handle if you have time.

**Comment by CheShengSu** [comment]
Will Apollo iOS Client support `Upload` in input types in the future?

**Comment by designatednerd** [comment]
@CheShengSu Meaning as a property on an input type rather than being at the root of the query?

**Comment by CheShengSu** [comment]
Yes. property `Upload` on an input type rather than being at the root of the query!

**Comment by designatednerd** [comment]
Got it! We're hoping to eventually but it's pretty low on the priority list. Honestly, uploads are sort of stapled on to GraphQL - we've found it often works better to use a purpose-built uploader then use a GraphQL Mutation to send up whatever URL you got back from the uploader.
